runtime/jcl/common/reflecthelp.c

```diff
@@ -2022,33 +2022,42 @@ permittedSubclassesHelper(JNIEnv *env, jobject cls)
 
 	permittedSubclassesCountPtr = getNumberOfPermittedSubclassesPtr(romClass);
 
-	stringArrayObject = (j9array_t) mmFuncs->J9AllocateIndexableObject(vmThread, stringArrayClass,
+	stringArrayObject = mmFuncs->J9AllocateIndexableObject(vmThread, stringArrayClass,
 				*permittedSubclassesCountPtr, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
 	if (NULL == stringArrayObject) {
 		goto heapoutofmemory;
 	}
 
-	result = vmFuncs->j9jni_createLocalRef(env, (j9object_t)stringArrayObject);
-
 	for (; index < *permittedSubclassesCountPtr; index++) {
 		J9UTF8* nameUTF = NULL;
 		j9object_t nameString = NULL;
 
 		nameUTF = permittedSubclassesNameAtIndex(permittedSubclassesCountPtr, index);
+
+		PUSH_OBJECT_IN_SPECIAL_FRAME(vmThread, stringArrayObject);
+
 		/* Translates string to a dot seperated name which is needed in Java code to get the class object. */
 		nameString = mmFuncs->j9gc_createJavaLangString(vmThread, J9UTF8_DATA(nameUTF), (U_32) J9UTF8_LENGTH(nameUTF), J9_STR_INTERN | J9_STR_XLAT);
+
+		stringArrayObject = POP_OBJECT_IN_SPECIAL_FRAME(vmThread);
+
 		if (NULL == nameString) {
 			goto heapoutofmemory;
 		}
 
 		J9JAVAARRAYOFOBJECT_STORE(vmThread, stringArrayObject, index, nameString);
 	}
-	goto done;
 
-heapoutofmemory:
-	vmFuncs->setHeapOutOfMemoryError(vmThread);
-	goto done;
+	result = vmFuncs->j9jni_createLocalRef(env, (j9object_t)stringArrayObject);
```

Comment:
```suggestion
	result = vmFuncs->j9jni_createLocalRef(env, stringArrayObject);
```
The cast is not required now.